src/OrchardCore.Modules/Orchard.Lucene/Controllers/AdminController.cs

```diff
@@ -173,33 +182,77 @@ public ActionResult Index()
             return RedirectToAction("Index");
         }
 
-        public async Task<IActionResult> Query(AdminQueryViewModel model)
+        public Task<IActionResult> Query(string indexName, string query, [FromServices] ITokenizer tokenizer)
+        {
+            query = String.IsNullOrWhiteSpace(query) ? "" : System.Text.Encoding.UTF8.GetString(Convert.FromBase64String(query));
+            return Query(new AdminQueryViewModel { IndexName = indexName, DecodedQuery = query }, tokenizer);
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Query(AdminQueryViewModel model, [FromServices] ITokenizer tokenizer)
         {
             if (!await _authorizationService.AuthorizeAsync(User, Permissions.ManageIndexes))
             {
                 return Unauthorized();
             }
 
-            if (!_luceneIndexProvider.Exists(model.IndexName))
+            model.Indices = _luceneIndexManager.List().ToArray();
+
+            // Can't query if there are no indices
+            if (model.Indices.Length == 0)
+            {
+                return RedirectToAction("Index");
+            }
+
+            if (String.IsNullOrEmpty(model.IndexName))
+            {
+                model.IndexName = model.Indices[0];
+            }
+
+            if (!_luceneIndexManager.Exists(model.IndexName))
             {
                 return NotFound();
             }
 
-            if (String.IsNullOrWhiteSpace(model.Query))
+            if (String.IsNullOrWhiteSpace(model.DecodedQuery))
             {
                 return View(model);
             }
 
+            if (String.IsNullOrEmpty(model.Parameters))
+            {
+                model.Parameters = "{ }";
+            }
+
             var luceneSettings = await _luceneIndexingService.GetLuceneSettingsAsync();
 
-            var queryParser = new QueryParser(LuceneSettings.DefaultVersion, "", new StandardAnalyzer(LuceneSettings.DefaultVersion));
-            var query = queryParser.Parse(model.Query);
+            var stopwatch = new Stopwatch();
+            stopwatch.Start();
 
-            _luceneIndexProvider.Search(model.IndexName, searcher =>
+            await _luceneIndexManager.SearchAsync(model.IndexName, searcher =>
             {
-                var docs = searcher.Search(query, 10);
-                model.Documents = docs.ScoreDocs.Select(hit => searcher.Doc(hit.Doc)).ToList();
-            });
+                var analyzer = _luceneAnalyzerManager.CreateAnalyzer("standardanalyzer");
+                var context = new LuceneQueryContext(searcher, LuceneSettings.DefaultVersion, analyzer);
+
+                var tokenizedContent = tokenizer.Tokenize(model.DecodedQuery, JObject.Parse(model.Parameters));
+
+                try
+                {
+                    var parameterizedQuery = JObject.Parse(tokenizedContent);
+                    var docs = _queryService.Search(context, parameterizedQuery);
```

Comment:
Lucene is not async and this service is specific to Lucene, but for the sake of future usage I am switching to async.